Static leak detector should print the "original" field in the test class that lead to ram usage tester failure

While porting Lucene tests to Java 9 module system, the static leak detector made some problems, because it tried to measure objects which are no longer reachable as encapsulated by the Java 9 module system.

But also in previous Java versions some classes referred to by static fields may not be measured, if a security manager is active and accessClassInPackage permission is not given.

Unfortunetely the error message on a `SecurityException` or `InaccessibleObjectException` (new in Java 9) does not really help to identify the static field in the test class. So it is hard to figure out which static field needs to be nulled.

I suggest to keep track on the field name in the test class before diving into the hierarchy behind a field. If during RAM usage measurement some exception occurs, the original field name should be printed as a new exception with the cause that lead to measurement error.

In addition, we should maybe add some automatic checks (based on code in forbiddenapis code that checks the module behind a class - we can only measure objects that are in `java.*` modules) that finds out if a class is encapsulated by Java 9 or hidden by security manager and don't measure it at all (just assigne some static "size" to instances without looking into them).

Here are some examples I found yesterday: https://github.com/apache/lucene-solr/commit/d0156b1126f094e4e469172d55842ed77cb82943

Stack trace was:
```
java.lang.IllegalStateException: Unable to access 'private final java.lang.reflect.Layer java.lang.reflect.Module.layer' to estimate memory usage
	at __randomizedtesting.SeedInfo.seed([1959EA728A13EC78]:0)
	at com.carrotsearch.randomizedtesting.rules.RamUsageEstimator.createCacheEntry(RamUsageEstimator.java:602)
	at com.carrotsearch.randomizedtesting.rules.RamUsageEstimator.measureSizeOf(RamUsageEstimator.java:545)
	at com.carrotsearch.randomizedtesting.rules.RamUsageEstimator.sizeOfAll(RamUsageEstimator.java:387)
	at com.carrotsearch.randomizedtesting.rules.StaticFieldsInvariantRule$1.afterAlways(StaticFieldsInvariantRule.java:127)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:43)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:367)
	at java.lang.Thread.run(java.base@9-ea/Thread.java:804)
Caused by: java.lang.reflect.InaccessibleObjectException: Cannot make a non-public member of class java.lang.reflect.Module accessible
	at sun.reflect.Reflection.throwInaccessibleObjectException(java.base@9-ea/Reflection.java:420)
	at java.lang.reflect.AccessibleObject.checkCanSetAccessible(java.base@9-ea/AccessibleObject.java:190)
	at java.lang.reflect.Field.checkCanSetAccessible(java.base@9-ea/Field.java:170)
	at java.lang.reflect.Field.setAccessible(java.base@9-ea/Field.java:164)
	at com.carrotsearch.randomizedtesting.rules.RamUsageEstimator$3.run(RamUsageEstimator.java:597)
	at com.carrotsearch.randomizedtesting.rules.RamUsageEstimator$3.run(RamUsageEstimator.java:594)
	at java.security.AccessController.doPrivileged(java.base@9-ea/Native Method)
	at com.carrotsearch.randomizedtesting.rules.RamUsageEstimator.createCacheEntry(RamUsageEstimator.java:594)
	... 13 more
```
If the leak detector would have printed those static fields as failing (actually SolrClients that referred to some internal JVM objects deep in hierrachy. The `layer` field does not really help.

Another one that also failed with Java 8 by @hossman:
https://issues.apache.org/jira/browse/LUCENE-7132?focusedCommentId=15222996&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15222996

The issue was a non-null IndexWriter. If the leak detector would have printed that the "writer" failed to be measured, it would have been easier to find this.

--- a/randomized-runner/src/main/java/com/carrotsearch/randomizedtesting/rules/StaticFieldsInvariantRule.java
+++ b/randomized-runner/src/main/java/com/carrotsearch/randomizedtesting/rules/StaticFieldsInvariantRule.java
@@ -124,7 +124,31 @@ public class StaticFieldsInvariantRule implements TestRule {
           }
         }
 
-        final long ramUsage = RamUsageEstimator.sizeOfAll(values);
+        final long ramUsage;
+        try {
+          ramUsage = RamUsageEstimator.sizeOfAll(values);
+        } catch (Exception ex) {
+          // some problem occurred while trying to measure (e.g. Java 9, SecurityManager).
+          // we iterate over all fields to get the "bad one":
+          final StringBuilder b = new StringBuilder();
+          b.append("Clean up static fields (in @AfterClass?) and null them, ")
+            .append("your test still has references to classes of which the ")
+            .append("sizes cannot be measured due to security restrictions or Java 9 ")
+            .append("module encapsulation:");
+          for (final Entry e : fieldsAndValues) {
+            try {
+              RamUsageEstimator.sizeOf(e.value);
+            } catch (Exception ex1) {
+              b.append("\n  - ").append(e.field);
+            }
+          }
+          
+          AssertionFailedError err = new AssertionFailedError(b.toString());
+          err.initCause(ex);
+          errors.add(err);
+          return;
+        }
+        
         if (ramUsage > leakThreshold) {
           // Count per-field information to get the heaviest fields.
           for (Entry e : fieldsAndValues) {

--- a/randomized-runner/src/test/java/com/carrotsearch/randomizedtesting/rules/TestStaticFieldsInvariantRule.java
+++ b/randomized-runner/src/test/java/com/carrotsearch/randomizedtesting/rules/TestStaticFieldsInvariantRule.java
@@ -1,5 +1,7 @@
 package com.carrotsearch.randomizedtesting.rules;
 
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.HashMap;
 import java.util.Map;
 
@@ -105,5 +107,42 @@ public class TestStaticFieldsInvariantRule extends WithNestedTestClass {
       .contains(".field2")
       .contains(".field3")
       .doesNotContain(".field5");
+  }
+  
+  static class Holder {
+    private final Path path;
+    
+    Holder() {
+      this.path = Paths.get(".");
+      final String name = this.path.getClass().getName();
+      RandomizedTest.assumeTrue(Path.class.getName() + " is not implemented by internal class in this JVM: " + name,
+        name.startsWith("sun.") || name.startsWith("jdk."));
+    }
+  }
+  
+  public static class FailsJava9 extends Base {
+    static Holder field0; 
+    
+    @BeforeClass
+    private static void setup() throws Exception {
+      field0 = new Holder();
+    }
+  }
+
+  @Test
+  public void testJava9Jigsaw() {
+    // check if we have Java 9 module system:
+    try {
+      Class.class.getMethod("getModule");
+    } catch (Exception e) {
+      RandomizedTest.assumeTrue("This test requires Java 9 module system (Jigsaw)", false);
+    }
+  
+    Result runClasses = JUnitCore.runClasses(FailsJava9.class);
+    Assertions.assertThat(runClasses.getFailures()).hasSize(1);
+    
+    Assertions.assertThat(runClasses.getFailures().get(0).getTrace())
+      .contains("sizes cannot be measured due to security restrictions or Java 9")
+      .contains(".field0");
   }
 }

